File: frontend/src/lib/clinic.ts
import { post, request } from "@/lib/api";

export type Address = {
  line1: string;
  line2: string;
  city: string;
  state: string;
  postal_code: string;
  country: string;
};

export type Clinic = {
  id: string;
  name: string;
  slug: string;
  phone: string | null;
  email: string | null;
  website: string | null;
  timezone: string;
  currency: string;
  status: "pending" | "active" | "suspended";
  address: Partial<Address> | null;
  onboarding_completed_at: string | null;
};

export type ClinicSettings = {
  consultation_duration_minutes: number;
  consultation_fee: string;
  follow_up_fee: string;
  follow_up_window_days: number;
  tax_percent: string;
  invoice_prefix: string;
  token_prefix: string;
};

export type Role = {
  id: string;
  slug: string;
  name: string;
  description: string;
};

export type StaffMember = {
  id: string;
  first_name: string;
  last_name: string;
  email: string;
  phone: string | null;
  status: string;
  role: Role | null;
  email_verified: boolean;
  last_login_at: string | null;
  is_you: boolean;
};

export type Invitation = {
  id: string;
  email: string;
  role: Role;
  invited_by_name: string;
  expires_at: string;
  created_at: string;
};

export type InvitationPreview = {
  clinic_name: string;
  role_name: string;
  email: string;
  first_name: string;
};

export const getClinic = () => request<Clinic>("/clinic");

export const saveClinic = (changes: Partial<Clinic> & { address?: Partial<Address> }) =>
  request<Clinic>("/clinic", { method: "PATCH", body: JSON.stringify(changes) });

export const getSettings = () => request<ClinicSettings>("/clinic/settings");

export const saveSettings = (changes: Partial<ClinicSettings>) =>
  request<ClinicSettings>("/clinic/settings", {
    method: "PATCH",
    body: JSON.stringify(changes),
  });

export const completeSetup = () => post<Clinic>("/clinic/complete-setup", {});

export const getRoles = () => request<Role[]>("/clinic/roles");

export const getStaff = () => request<StaffMember[]>("/staff");

export const getInvitations = () => request<Invitation[]>("/staff/invitations");

export const sendInvitation = (payload: {
  email: string;
  first_name: string;
  last_name: string;
  role_slug: string;
}) => post<{ acknowledged: boolean; email_configured: boolean }>("/staff/invitations", payload);

export const revokeInvitation = (id: string) =>
  request<{ acknowledged: boolean }>(`/staff/invitations/${id}`, { method: "DELETE" });

export const changeRole = (memberId: string, roleSlug: string) =>
  request<{ acknowledged: boolean }>(`/staff/${memberId}/role`, {
    method: "PATCH",
    body: JSON.stringify({ role_slug: roleSlug }),
  });

export const suspendMember = (memberId: string) =>
  post<{ acknowledged: boolean }>(`/staff/${memberId}/suspend`, {});

export const restoreMember = (memberId: string) =>
  post<{ acknowledged: boolean }>(`/staff/${memberId}/restore`, {});

export const previewInvitation = (token: string) =>
  request<InvitationPreview>(`/invitations/${encodeURIComponent(token)}`);

export const acceptInvitation = (token: string, password: string) =>
  post<{ acknowledged: boolean }>("/invitations/accept", { token, password });

/** A clinic's own currency, written the way its staff would expect to read it. */
export function money(amount: string, currency: string, locale = "en-IN"): string {
  const value = Number(amount);
  if (!Number.isFinite(value)) return amount;
  return new Intl.NumberFormat(locale, {
    style: "currency",
    currency,
    maximumFractionDigits: 2,
  }).format(value);
}

File: backend/app/schemas/clinic.py
"""Request and response shapes for clinic setup and staff."""

from __future__ import annotations

import datetime as dt
import re
import uuid
from decimal import Decimal
from typing import Annotated, Any

from pydantic import BaseModel, ConfigDict, EmailStr, Field, field_validator

# Money crosses the wire as a string. A JSON number is a double, and a double
# cannot hold 0.1 + 0.2, which is not a property to give a clinic's fees.
# Two digits for the state, the PAN, then an entity number, a Z and a check
# character.
GSTIN = re.compile(r"^[0-9]{2}[A-Z]{5}[0-9]{4}[A-Z][1-9A-Z]Z[0-9A-Z]$")

PREFIX = re.compile(r"^[A-Z0-9-]+$")

Money = Annotated[Decimal, Field(max_digits=12, decimal_places=2, ge=0)]


class _Trimmed(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)


class Address(_Trimmed):
    line1: str = Field(default="", max_length=160)
    line2: str = Field(default="", max_length=160)
    city: str = Field(default="", max_length=80)
    state: str = Field(default="", max_length=80)
    postal_code: str = Field(default="", max_length=16)
    country: str = Field(default="India", max_length=80)


class ClinicUpdate(_Trimmed):
    name: str | None = Field(default=None, min_length=2, max_length=160)
    phone: str | None = Field(default=None, max_length=32)
    email: EmailStr | None = None
    website: str | None = Field(default=None, max_length=255)
    timezone: str | None = Field(default=None, max_length=64)
    currency: str | None = Field(default=None, min_length=3, max_length=3)
    address: Address | None = None

    @field_validator("currency")
    @classmethod
    def _upper(cls, value: str | None) -> str | None:
        return value.upper() if value else value


class ClinicSettingsUpdate(_Trimmed):
    consultation_duration_minutes: int | None = Field(default=None, ge=5, le=240)
    consultation_fee: Money | None = None
    follow_up_fee: Money | None = None
    follow_up_window_days: int | None = Field(default=None, ge=0, le=365)
    tax_percent: Annotated[Decimal, Field(ge=0, le=100, decimal_places=2)] | None = None
    # Printed at the front of every bill number, as in INV/2026-27/0001.
    invoice_prefix: str | None = Field(default=None, min_length=1, max_length=8)
    token_prefix: str | None = Field(default=None, min_length=1, max_length=4)
    # Printed on bills when the clinic is registered for GST. Sent empty, it
    # comes off.
    gstin: str | None = Field(default=None, max_length=15)

    @field_validator("invoice_prefix", "gstin")
    @classmethod
    def _capitals(cls, value: str | None) -> str | None:
        return value.upper() if value else value

    @field_validator("invoice_prefix")
    @classmethod
    def _prefix(cls, value: str | None) -> str | None:
        if value and not PREFIX.fullmatch(value):
            raise ValueError("Use letters, digits and dashes only, like INV or LC-1.")
        return value

    @field_validator("gstin")
    @classmethod
    def _gstin(cls, value: str | None) -> str | None:
        if value and not GSTIN.fullmatch(value):
            raise ValueError("A GSTIN is 15 characters, like 27ABCDE1234F1Z5.")
        return value


class ClinicSettingsOut(BaseModel):
    consultation_duration_minutes: int
    consultation_fee: str
    follow_up_fee: str
    follow_up_window_days: int
    tax_percent: str
    invoice_prefix: str
    token_prefix: str
    gstin: str


class ClinicOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    slug: str
    phone: str | None
    email: str | None
    website: str | None
    timezone: str
    currency: str
    status: str
    address: dict[str, Any] | None
    onboarding_completed_at: dt.datetime | None


class RoleOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    slug: str
    name: str
    description: str


class StaffMemberOut(BaseModel):
    id: uuid.UUID
    first_name: str
    last_name: str
    email: str
    phone: str | None
    status: str
    role: RoleOut | None
    email_verified: bool
    last_login_at: dt.datetime | None
    is_you: bool


class InviteRequest(_Trimmed):
    email: EmailStr
    first_name: str = Field(min_length=1, max_length=80)
    last_name: str = Field(min_length=1, max_length=80)
    role_slug: str = Field(min_length=1, max_length=48)

    @field_validator("email")
    @classmethod
    def _lower(cls, value: str) -> str:
        return value.lower()


class InvitationOut(BaseModel):
    id: uuid.UUID
    email: str
    role: RoleOut
    invited_by_name: str
    expires_at: dt.datetime
    created_at: dt.datetime


class InvitationPreview(BaseModel):
    """What someone opening an invitation link is shown before they accept.

    Only the clinic and the role. Nothing about who else works there, because
    whoever holds this link has not proved they are the intended recipient.
    """

    clinic_name: str
    role_name: str
    email: str
    first_name: str


class AcceptInvitation(_Trimmed):
    token: str = Field(min_length=1, max_length=256)
    password: str = Field(min_length=1, max_length=128)


class ChangeRoleRequest(_Trimmed):
    role_slug: str = Field(min_length=1, max_length=48)
